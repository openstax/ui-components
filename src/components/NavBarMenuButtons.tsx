import React from "react";
import {
  Dialog,
  DialogTrigger,
  Menu,
  MenuItem,
  MenuTrigger,
  Popover,
  PopoverProps,
} from "react-aria-components";
import styled from "styled-components";
import { colors } from "../../src/theme";
import { NavBarButton, NavBarButtonProps } from "./NavBarButton";

export const NavBarMenuItem = styled(MenuItem)``;

export const PopoverContainer = styled.div`
  padding: 1.6rem;
`;

const NavBarPopover = styled(Popover)`
  margin-top: -1rem;
  border-top: 0.4rem solid ${colors.palette.darkGreen};
  box-shadow: 0 0.4rem 0.4rem 0 #00000033;
  background: #fff;

  ${NavBarMenuItem} {
    font-size: 1.4rem;
    min-height: 4rem;
    padding: 0 1.6rem;
    cursor: pointer;
    display: flex;
    align-items: center;

    &:hover {
      background: ${colors.palette.neutralLighter};
    }

    &:active {
      font-weight: bold;
    }
  }
`;

export type NavBarItemProps = React.PropsWithChildren<{
  label?: string;
  icon?: string | React.ReactNode;
  buttonProps?: NavBarButtonProps;
  popoverProps?: PopoverProps;
  className?: string;
}> & (
  | { label: string; icon?: React.ReactNode; "aria-label"?: string }
  | { label?: never; icon: React.ReactNode; "aria-label": string }
);

const NavBarBaseButton = ({
  isMenu,
  children,
  className,
  label,
  icon,
  "aria-label": ariaLabel,
  ...props
}: NavBarItemProps & {
  isMenu: boolean;
}) => {
  const Trigger = isMenu ? MenuTrigger : DialogTrigger;
  const Content = isMenu ? Menu : Dialog;

  return (
    <Trigger>
      <NavBarButton
        className={className}
        aria-label={ariaLabel || label}
        label={label}
        icon={icon}
        {...props.buttonProps}
      />
      <NavBarPopover {...props.popoverProps}>
        <Content>{children}</Content>
      </NavBarPopover>
    </Trigger>
  );
};

export const NavBarPopoverButton = (props: NavBarItemProps) => (
  <NavBarBaseButton {...props} isMenu={false} />
);

export const NavBarMenuButton = (props: NavBarItemProps) => (
  <NavBarBaseButton {...props} isMenu={true} />
);
