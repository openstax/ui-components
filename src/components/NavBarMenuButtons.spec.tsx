import { Menu, MenuItem } from "react-aria-components";
import renderer from "react-test-renderer";
import { NavBarMenuButton, NavBarPopoverButton } from "./NavBarMenuButtons";

describe("NavBarPopoverButton", () => {
  it("matches snapshot", () => {
    const tree = renderer
      .create(
        <NavBarPopoverButton label="Test menu">
          Popover content
        </NavBarPopoverButton>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("matches custom aria label snapshot", () => {
    const tree = renderer
      .create(
        <NavBarPopoverButton label="Test menu" aria-label="Custom label">
          Popover content
        </NavBarPopoverButton>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("NavBarMenuButton", () => {
  it("matches snapshot", () => {
    const tree = renderer
      .create(
        <NavBarMenuButton label="Test menu">
          <Menu>
            <MenuItem>Menu item</MenuItem>
            <MenuItem>Another menu item</MenuItem>
          </Menu>
        </NavBarMenuButton>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
