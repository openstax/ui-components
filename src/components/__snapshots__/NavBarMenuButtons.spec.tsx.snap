// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBarMenuButton matches snapshot 1`] = `
<button
  aria-expanded={false}
  aria-haspopup={true}
  className="sc-bczRLJ dKgZvy"
  data-rac=""
  data-react-aria-pressable={true}
  id="react-aria-7"
  onBlur={[Function]}
  onClick={[Function]}
  onDragStart={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span>
    Test menu
  </span>
</button>
`;

exports[`NavBarPopoverButton matches custom aria label snapshot 1`] = `
<button
  aria-expanded={false}
  aria-label="Custom label"
  className="sc-bczRLJ dKgZvy"
  data-rac=""
  data-react-aria-pressable={true}
  id="react-aria-5"
  onBlur={[Function]}
  onClick={[Function]}
  onDragStart={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span>
    Test menu
  </span>
</button>
`;

exports[`NavBarPopoverButton matches snapshot 1`] = `
<button
  aria-expanded={false}
  className="sc-bczRLJ dKgZvy"
  data-rac=""
  data-react-aria-pressable={true}
  id="react-aria-2"
  onBlur={[Function]}
  onClick={[Function]}
  onDragStart={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span>
    Test menu
  </span>
</button>
`;
